silx/io/h5py_utils.py

```diff
@@ -0,0 +1,317 @@
+    """Yield an HDF5 dataset or group (retry until it can be instantiated).
+
+    :param str filename:
+    :param bool retry_invalid: retry when item is missing or not valid
+    :param callable or None validate:
+    :yields Dataset, Group or None:
+    """
+    with File(filename) as h5file:
+        try:
+            item = h5file[name]
+        except KeyError as e:
+            if "doesn't exist" in str(e):
+                if retry_invalid:
+                    raise retry_mod.RetryError
+                else:
+                    item = None
+            else:
+                raise
+        if callable(validate) and item is not None:
+            if not validate(item):
+                if retry_invalid:
+                    raise retry_mod.RetryError
+                else:
+                    item = None
+        yield item
+
+
+def _top_level_names(filename, include_only=group_has_end_time):
+    """Return all valid top-level HDF5 names.
+
+    :param str filename:
+    :param callable or None include_only:
+    :returns list(str):
+    """
+    with File(filename) as h5file:
+        try:
+            if callable(include_only):
+                return [name for name in h5file["/"] if include_only(h5file[name])]
+            else:
+                return list(h5file["/"])
+        except KeyError:
+            raise retry_mod.RetryError
+
+
+top_level_names = retry()(_top_level_names)
+safe_top_level_names = retry_in_subprocess()(_top_level_names)
+
+
+class File(h5py.File):
+    """Takes care of HDF5 file locking and SWMR mode without the need
+    to handle those explicitely.
+
+    When using this class, you cannot open different files simultatiously
+    with different modes because the locking flag is an environment variable.
+    """
+
+    _HDF5_FILE_LOCKING = None
+    _NOPEN = 0
+    _SWMR_LIBVER = "latest"
+
+    def __init__(
+        self,
+        filename,
+        mode=None,
+        enable_file_locking=None,
+        swmr=None,
+        libver=None,
+        **kwargs
+    ):
+        """The arguments `enable_file_locking` and `swmr` should not be
+        specified explicitly for normal use cases.
+
+        :param str filename:
+        :param str or None mode: read-only by default
+        :param bool or None enable_file_locking: by default it is disabled for `mode='r'`
+                                                 and `swmr=False` and enabled for all
+                                                 other modes.
+        :param bool or None swmr: try both modes when `mode='r'` and `swmr=None`
+        :param **kwargs: see `h5py.File.__init__`
+        """
+        if mode is None:
+            mode = "r"
+        elif mode not in ("r", "w", "w-", "x", "a", "r+"):
+            raise ValueError("invalid mode {}".format(mode))
+        if not HAS_SWMR:
+            swmr = False
+
+        if enable_file_locking is None:
+            enable_file_locking = bool(mode != "r" or swmr)
```

Comment:
"r+" is like append mode in most regards so file locking need to be enabled. If you have any case that does this, make sure the change that asap, you'll be corrupting the HDF5 file otherwise.

Reply:
My bad sorry, I thought it was a mode "create" file if does not exist and read only (that you couldn't do any "append" on it but I was wrong)